core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java

```diff
@@ -308,12 +308,20 @@ private static RexNode convertNvl(SqlRexContext cx, SqlCall call) {
         cx.convertExpression(call.getOperandList().get(1));
     final RelDataType type =
         cx.getValidator().getValidatedNodeType(call);
+    // Preserve Operand Nullability
     return rexBuilder.makeCall(type, SqlStdOperatorTable.CASE,
         ImmutableList.of(
             rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_NULL,
                 operand0),
-            rexBuilder.makeCast(type, operand0),
-            rexBuilder.makeCast(type, operand1)));
+            rexBuilder.makeCast(
+                cx.getTypeFactory()
+                    .createTypeWithNullability(type, operand0.getType().isNullable()),
+                operand0),
+            rexBuilder.makeCast(
+                cx.getTypeFactory()
```

Comment:
I agree that operand1 should keep its nullability. But operand0 would be not nullable after rewrite.

Reply:
@chunweilei 
1) Consider If that operand depends on an input reference(which is nullable)
2) If in future that operand pushed down like in [ProjectJoinTransposeRule](https://github.com/apache/calcite/blob/bebe473fab2e242736614659ed6e5d04eeeb8bf5/core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java), there will be an issue as the operand and its reference have different nullability.
3) The output of the NVL operation is Not Null but there is no need to change the nullability of the operands as in calcite rules we might push the operands down in some rules causing errors ( operand referenced input type is nullable whereas we forced the operand to be not Null)
Do you think hence it makes sense to let the operand nullability what it was as the output is anyway not nullable?

Reply:
@chunweilei NVL(null, null) is valid sql.

Reply:
The return type should also be nullable if both inputs are nullable.

Reply:
`
select nvl(“name”, “name”) FROM “hr”.“emps”;
`
For this query the output type was indeed nullable (both names are nullable) so doesn't seem like an issue.

Reply:
Got it! Thank you for your explanation.